husbandry.lua

```diff
@@ -0,0 +1,326 @@
+
+local function canMilkCreature(u)
+    if dfhack.units.isMilkable(u) and not dfhack.units.isPet(u) then
+        local mt_milk = dfhack.units.getMiscTrait(u, df.misc_trait_type.MilkCounter, false)
+        if not mt_milk then return true else return false end
+    else
+        return nil
+    end
+end
+
+---@param p1 df.coord
+---@param p2 df.coord
+---@return number
+function distance(p1, p2)
+    return math.max(math.abs(p1.x - p2.x), math.abs(p1.y - p2.y)) + 2 * math.abs(p1.z - p2.z)
+end
+
+---find appropriate workshop to milk or shear an animal
+---@param unit df.unit
+---@param collection table<integer,df.building_workshopst>
+---@return df.building_workshopst?
+local function getAppropriateWorkshop(unit, collection)
+    local zone_ref = dfhack.units.getGeneralRef(unit, df.general_ref_type.BUILDING_CIVZONE_ASSIGNED)
+    local zone = zone_ref and zone_ref:getBuilding() or nil
+
+    -- if animal is assigned to a zone containing workshops, only use those
+    if zone then
+        local contains_workshop = false
+        local best = nil
+        local worst_load = 10
+        for _, workshop in pairs(collection[zone.z] or {}) do
+            if dfhack.buildings.containsTile(zone, workshop.centerx, workshop.centery) then
+                contains_workshop = true
+                local workshop_pos = xyz2pos(workshop.centerx, workshop.centery, workshop.z)
+                if dfhack.maps.canWalkBetween(unit.pos, workshop_pos) and #workshop.jobs < worst_load then
+                    worst_load = #workshop.jobs
+                    best = workshop
+                end
+            end
+        end
+        if contains_workshop or state.pasture then
+            return best
+        end
+    elseif not state.roaming then
+        return nil -- not treating roaming animals
+    end
+    -- otherwise, use the closest workshop to the animal
+    local closest = nil
+    local dist = nil
+    for _, level in pairs(collection) do
+        for _, workshop in pairs(level) do
+            local workshop_pos = xyz2pos(workshop.centerx, workshop.centery, workshop.z)
+            if dfhack.maps.canWalkBetween(unit.pos, workshop_pos) then
+                local d = distance(unit.pos, workshop_pos)
+                if not closest or d < dist then
+                    closest = workshop
+                    dist = d
+                end
+            end
+        end
+    end
+    return #closest.jobs < 10 and closest or nil
+end
+
+local function shearCreature(unit, workshop)
+    local job = ic.make_job()
+    job.job_type = df.job_type.ShearCreature
+    dfhack.job.addGeneralRef(job, df.general_ref_type.UNIT_SHEAREE, unit.id)
+    ic.assignToWorkshop(job, workshop)
+end
+
+local function milkCreature(unit, workshop)
+    local job = ic.make_job()
+    job.job_type = df.job_type.MilkCreature
+    dfhack.job.addGeneralRef(job, df.general_ref_type.UNIT_MILKEE, unit.id)
+    ic.assignToWorkshop(job, workshop)
+end
```

Comment:
this to me indicates that the functionality that's here being used from idle-crafting should at least be considered for being moved to a library

Reply:
Those are part of the functions that were recently merged into the library. I can fix this PR, but I can also do a follow up where I replace these functions in all the related tools.

Reply:
A follow up is fine. This code is fine as it stands and I see no reason to delay merging it so we can include it in the next pre-release for UAT